edgehog-device-runtime-forwarder/src/connections.rs

```diff
@@ -0,0 +1,362 @@
+    pub(crate) fn handle_proto_msg(&mut self, proto_msg: ProtoMessage) -> Result<(), Error> {
+        // remove from the collection all the terminated connections
+        self.connections.remove_terminated_connections();
+
+        // handle only HTTP requests, not other kind of protobuf messages
+        let http_req = match proto_msg.protocol {
+            ProtoProtocol::Http(Http::Request(http_req)) => http_req,
+            ProtoProtocol::Http(Http::Response(_http_res)) => {
+                error!("Http response should not be sent by the bridge");
+                return Err(Error::WrongMessage);
+            }
+            ProtoProtocol::WebSocket(_ws) => {
+                error!("WebSocket messages are not supported yet");
+                return Err(Error::WrongMessage);
+            }
+        };
+
+        self.connections.add_http(http_req)?;
+
+        Ok(())
+    }
+
+    /// Try to establish again a WebSocket connection with the bridge in case the connection is lost.
+    #[instrument(skip_all)]
+    pub(crate) async fn reconnect(&mut self) -> Result<(), Error> {
+        debug!("trying to reconnect");
+
+        self.ws_stream = Self::ws_connect(&self.url).await?;
+
+        info!("reconnected");
+
+        Ok(())
+    }
+
+    /// Close all the connections the device has established (e.g., with TTYD).
+    #[instrument(skip_all)]
+    pub(crate) fn disconnect(&mut self) {
+        info!("closing all the connections");
+        self.connections.disconnect();
+    }
+}
+
+/// Possible events happening on a WebSocket connection.
+pub(crate) enum WebSocketEvents {
+    Receive(Result<TungMessage, TungError>),
+    Send(ProtoMessage),
+    Ping,
+}
+
+/// Collection of connections between the device and the bridge.
+#[derive(Debug)]
+pub(crate) struct Connections {
+    /// Collection mapping every Connection ID with the corresponding [`tokio task`](tokio::task) spawned to
+    /// handle it.
+    connections: HashMap<Id, ConnectionHandle>,
+    /// Write side of the channel used by each connection to send data to the [`ConnectionsManager`].
+    /// This field is only cloned and passed to every connection when created.
+    tx_ws: UnboundedSender<ProtoMessage>,
+}
+
+impl Connections {
+    /// Initialize the Connections' collection.
+    pub(crate) fn init(tx_ws: UnboundedSender<ProtoMessage>) -> Self {
+        Self {
+            connections: HashMap::new(),
+            tx_ws,
+        }
+    }
+
+    /// Create a new [`Connection`] in case a new HTTP request is received.
+    #[instrument(skip(self, http_req))]
+    pub(crate) fn add_http(&mut self, http_req: HttpRequest) -> Result<(), Error> {
+        let id = http_req.request_id.clone();
+
+        // check if there exist a connection with that id
+        let Entry::Vacant(entry) = self.connections.entry(id.clone()) else {
+            error!("entry already occupied");
+            return Err(Error::IdAlreadyUsed(id));
+        };
```

Comment:
Intead of on every message to check if all the entries are finished. Create a method like `connection` that will return an entry only if is not finished and will remove the finished one.

Reply:
Also move the remove_terminated_connections into a function `add_connection` that will cleanup all the hashmap.